File: scala/src/main/scala/nonamedb/api/Service.scala
package nonamedb.api

import akka.actor.ActorSystem
import akka.http.scaladsl.Http
import akka.http.scaladsl.model._
import akka.http.scaladsl.server.Directives._
import akka.http.scaladsl.server.Route
import akka.util.ByteString
import nonamedb.storage.Engine

import scala.concurrent.Future

class Service(engine: Engine)(implicit system: ActorSystem) {

  import system.dispatcher

  def start(bindAddress: String, bindPort: Int): Future[Http.ServerBinding] =
    Http().newServerAt(bindAddress, bindPort).bind(routes)

  val routes: Route =
    path(Segment) { id =>
      concat(
        get {
          onComplete(engine.get(id)) { result =>
            result.toOption.flatten match {
              case Some(data) =>
                complete(
                  HttpResponse(
                    entity = HttpEntity(MediaTypes.`application/octet-stream`, data)
                  )
                )
              case None =>
                reject
            }
          }
        },
        (post | put) {
          extractDataBytes { dataStream =>
            complete {
              dataStream
                .runFold(ByteString.empty) {
                  case (acc, current) =>
                    acc ++ current
                }
                .flatMap { data =>
                  engine.put(id, data.toArray)
                }
            }
          }
        },
        delete {
          complete {
            engine.put(id, Array.emptyByteArray)
          }
        }
      )
    }
}

File: clojure/src/nonamedb/api/service.clj
(ns nonamedb.api.service
  (:use [compojure core])
  (:require [compojure.route :as route]
            [compojure.handler :as handler]
            [nonamedb.storage.engines.memory-engine :refer [get-from put-into]]))

(defn create-service [engine]
  (handler/site
    (defroutes
      api-routes

      (GET "/:key" [key]
        (when-let [result (get-from engine key)]
          {:status       200
           :content-type "application/octet-stream"
           :body         (String. result)}))

      (POST "/:key" [key :as request]
        (if-let [result (put-into engine key (byte-array (map byte (slurp (:body request)))))]
          {:status 200}
          {:status       500
           :content-type "text/plain"
           :body         "Failed to update data"}))

      (PUT "/:key" [key :as request]
        (if-let [result (put-into engine key (byte-array (map byte (slurp (:body request)))))]
          {:status 200}
          {:status       500
           :content-type "text/plain"
           :body         "Failed to update data"}))


      (DELETE "/:key" [key]
        (if-let [result (put-into engine key (byte-array []))]
          {:status 200}
          {:status       500
           :content-type "text/plain"
           :body         "Failed to delete data"}))

      (route/not-found nil))))
